File: python/netflix-rpc.py
"""
script to display currently watching netflixshow in discord-rpc
"""

import requests
import time
from pypresence import Presence

class assets:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    ENDC = '\033[0m'

try:
    with open("config.csv","r") as f:
        f = f.read().split("\n")
        f = f[1].split("|")
        shakti_ver = f[0]
        cookie = f[1]
        print(f"[{assets.OKGREEN}+{assets.ENDC}][{str(time.asctime()).split(' ')[-2]}] <cookie> and <api_version> extracted from config <{shakti_ver}>,<{cookie[0:10]}[...]>\n")
except:
	print("config.csv not found...")
	print("creating config.csv file...")
	print("requesting newest api version from /xNaCly/netflix-rpc/main/shakti")
	shakti = requests.get("https://raw.githubusercontent.com/xNaCly/netflix-rpc/main/shakti").text
	with open("config.csv","w") as f:
		f.write(f"shaktivver|cookie\n{shakti}|COOKIEPLACEHOLDER")
	print("created config.csv with newest api version, please replace <COOKIEPLACEHOLDER> with your cookie!")
	exit()


timer = int(time.time())
start1 = time.time()
RPC = Presence(740305744116580382)
RPC.connect()
print(f"[{assets.OKGREEN}+{assets.ENDC}][{str(time.asctime()).split(' ')[-2]}] connected to discord... [{round(time.time()-start1,2)}ms]")

headers={
        "cookie": cookie,
}

start = time.time()
info = requests.get(f"https://www.netflix.com/api/shakti/{shakti_ver}/profiles", headers=headers).json()["profiles"]
profiles = []

for profile in info:
    profiles.append(profile["firstName"].upper())

print(f"[{assets.OKGREEN}+{assets.ENDC}][{str(time.asctime()).split(' ')[-2]}] connected to netflix... [{round(time.time()-start,2)}ms]")
print(f"[:][{str(time.asctime()).split(' ')[-2]}] following profiles detected: ({', '.join(profiles)})\n")

while True:
    start = time.time()
    r = requests.get(f"https://www.netflix.com/api/shakti/{shakti_ver}/viewingactivity/", headers=headers).json()
    lastviewed = r["viewedItems"][0]
    keys = list(lastviewed.keys())

    if "seriesTitle" in keys:
        print(f"[{assets.OKGREEN}+{assets.ENDC}][{str(time.asctime()).split(' ')[-2]}] https://www.netflix.com/api/shakti/{shakti_ver}/viewingactivity/ {lastviewed['seriesTitle']}({lastviewed['title']})[{str(round(lastviewed['bookmark']/60))}/{str(round(lastviewed['duration']/60))}min] [{round(time.time()-start,2)}ms]")
        RPC.update(
            state=lastviewed["seriesTitle"],
            start=timer,
            large_image="logo",
            large_text=lastviewed["title"],
            small_image="logo",
            small_text=f"Progress: {str(round(lastviewed['bookmark']/60))}/{str(round(lastviewed['duration']/60))}min"
        )
    else:
        print(f"[{assets.OKGREEN}+{assets.ENDC}][{str(time.asctime()).split(' ')[-2]}] https://www.netflix.com/api/shakti/{shakti_ver}/viewingactivity/ {lastviewed['title']}({lastviewed['videoTitle']})[{str(round(lastviewed['bookmark']/60))}/{str(round(lastviewed['duration']/60))}min] [{round(time.time()-start,2)}ms]")
        RPC.update(
            state=lastviewed["title"],
            start=timer,
            large_image="logo",
            large_text=lastviewed["videoTitle"],
            small_image="logo",
            small_text=f"Progress: {str(round(lastviewed['bookmark']/60))}/{str(round(lastviewed['duration']/60))}min"
        )

    time.sleep(30)

File: javascript/netflix-rpc.js
const { Client } = require("discord-rpc");
const fetch = require("node-fetch");
const { cookie } = require("./config.json");

function logger(content) {
	console.log(`[${new Date().toTimeString().split(" ")[0]}]`, content);
}

var Timestamp = Date.now();
var startTimestamp = new Date();

const sleep = (ms) => new Promise((res) => setTimeout(res, ms));

async function reconnect() {
	logger(`initiating ${__filename}`);
	logger("<cookie> and <api_version> extracted from config");
	logger("-----------------------------------------------");
	try {
		await login();
	} catch (error) {
		console.trace(error);
		if (!cookie || cookie == "COOKIEPLACEHOLDER") {
			throw "Cookie missing,\nplease go to https://github.com/xNaCly/netflix-rpc/tree/master/javascript and follow the steps.\n";
		}
		setTimeout(async () => {
			await reconnect();
		}, 1000 * 10);
	}
	logger("-----------------------------------------------");
}

reconnect();

var api_ver;

async function send_req() {
	api_ver = await fetch("https://raw.githubusercontent.com/xNaCly/netflix-rpc/master/shakti");
	api_ver = await api_ver.text();
	api_ver = api_ver.trim();
}

async function get_profiles() {
	let profiles = await fetch(`https://www.netflix.com/api/shakti/${api_ver}/profiles/`, {
		headers: {
			"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:86.0) Gecko/20100101 Firefox/86.0",
			Accept: "*/*",
			"Accept-Language": "en-US,en;q=0.5",
			"Accept-Encoding": "gzip, deflate, br",
			DNT: 1,
			Connection: "keep-alive",
			Cookie: cookie,
			"Sec-GPC": 1,
			"Cache-Control": "max-age=0",
		},
		method: "GET",
		mode: "cors",
	});
	profiles = await profiles.json();
	return profiles.profiles.map((x) => x.firstName.toUpperCase());
}

async function get_history_last() {
	let history = await fetch(`https://www.netflix.com/api/shakti/${api_ver}/viewingactivity/`, {
		headers: {
			"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:86.0) Gecko/20100101 Firefox/86.0",
			Accept: "*/*",
			"Accept-Language": "en-US,en;q=0.5",
			"Accept-Encoding": "gzip, deflate, br",
			DNT: 1,
			Connection: "keep-alive",
			Cookie: cookie,
			"Sec-GPC": 1,
			"Cache-Control": "max-age=0",
		},
		method: "GET",
		mode: "cors",
	});
	history = await history.json();
	return history.viewedItems[0];
}

async function login() {
	logger(`logging into discord...`);
	const rpc = new Client({
		transport: "ipc",
	});

	logger(`connected to discord... [${(Date.now() - Timestamp) / 1000}ms]`);
	logger("-----------------------------------------------");
	Timestamp = Date.now();
	logger(`logging into netflix...`);
	await send_req();
	const profiles = await get_profiles();
	logger(`connected to netflix... [${(Date.now() - Timestamp) / 1000}ms]`);
	logger(`following profiles detected: ${profiles.join(", ")}`);

	item = await get_history_last();

	rpc.on("ready", async () => {
		while (true) {
			logger(`updating rpc: api/shakti/${api_ver}/viewingactivity/ ...`);
			if (Object.keys(item).includes("seriesTitle")) {
				rpc.setActivity({
					state: item.seriesTitle,
					startTimestamp,
					largeImageKey: "logo",
					largeImageText: item.title,
					smallImageKey: "logo",
					smallImageText: `Progress: ${Math.round(item.bookmark / 60)}/${Math.round(item.duration / 60)}`,
					instance: false,
				});
			} else {
				rpc.setActivity({
					state: item.title,
					startTimestamp,
					largeImageKey: "logo",
					largeImageText: item.videoTitle,
					smallImageKey: "logo",
					smallImageText: `Progress: ${Math.round(item.bookmark / 60)}/${Math.round(item.duration / 60)}`,
					instance: false,
				});
			}

			item = await get_history_last();
			await sleep(15 * 1000);
		}
	});

	rpc.on("disconnected", async () => {
		console.log("disconnected");
		await reconnect();
	});

	rpc.login({ clientId: "740305744116580382" });
	return;
}
